File: go/500.600/515.FindLargestValueInEachTreeRow.go
package main

import (
	"fmt"
	"math"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func largestValues(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	queue := []*TreeNode{root}
	ans := []int{}
	for len(queue) > 0 {
		mx := math.MinInt
		qu := []*TreeNode{}
		for _, curr := range queue {
			mx = max(mx, curr.Val)
			if curr.Left != nil {
				qu = append(qu, curr.Left)
			}
			if curr.Right != nil {
				qu = append(qu, curr.Right)
			}
		}
		queue = qu
		ans = append(ans, mx)
	}
	return ans
}

func max(i, j int) int {
	if i > j {
		return i
	}
	return j
}

func main() {
	head := &TreeNode{Val: 1}
	head.Left = &TreeNode{Val: 3}
	head.Left.Left = &TreeNode{Val: 5}
	head.Left.Right = &TreeNode{Val: 3}
	head.Right = &TreeNode{Val: 2}
	head.Right.Right = &TreeNode{Val: 9}
	fmt.Println(largestValues(head), "[1,3,9]")

	head = &TreeNode{Val: 1}
	head.Left = &TreeNode{Val: 2}
	head.Right = &TreeNode{Val: 3}
	fmt.Println(largestValues(head), "[1,3]")
}

File: 500.600/515.FindLargestValueInEachTreeRow.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

struct TreeNode
{
    int val{};
    TreeNode *left{};
    TreeNode *right{};
    TreeNode() {}
    TreeNode(int x) : val(x) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class FindLargestValueInEachTreeRow
{
public:
    vector<int> largestValues(TreeNode *root)
    {
        vector<int> rets;
        if (!root)
        {
            return rets;
        }
        queue<TreeNode *> qu;
        qu.push(root);
        while (!qu.empty())
        {
            int mx = INT_MIN;
            int size = qu.size();
            for (int i = 0; i < size; ++i)
            {
                TreeNode *curr = qu.front();
                qu.pop();
                mx = max(mx, curr->val);
                if (curr->left)
                {
                    qu.push(curr->left);
                }
                if (curr->right)
                {
                    qu.push(curr->right);
                }
            }
            rets.push_back(mx);
        }
        return rets;
    }
};

void print(const vector<int> &ret)
{
    cout << "[";
    for (int x : ret)
    {
        cout << x << ",";
    }
    cout << "]" << endl;
}

void del(TreeNode *h)
{
    if (h->left)
    {
        del(h->left);
    }
    if (h->right)
    {
        del(h->right);
    }
    delete h;
}

int main(int argc, char const *argv[])
{
    FindLargestValueInEachTreeRow cls;
    TreeNode *head = new TreeNode(1);
    head->left = new TreeNode(3);
    head->left->left = new TreeNode(5);
    head->left->right = new TreeNode(3);
    head->right = new TreeNode(2);
    head->right->right = new TreeNode(9);
    //[1,3,9]
    print(cls.largestValues(head));
    del(head);

    head = new TreeNode(1);
    head->left = new TreeNode(2);
    head->right = new TreeNode(3);
    //[1,3]
    print(cls.largestValues(head));
    del(head);
    return 0;
}
